### PR TITLE
chore: remove temporary allocations from `num_bits`

### LINKED ISSUE #4759: Update `FieldElement.num_bits()` to return 1 for a zero field element.
`FieldElement.num_bits()` currently returns 0 for a zero field element:
https://github.com/noir-lang/acvm/blob/39b8a41293e567971f700f61103852cb987a8d16/acir_field/src/generic_ark.rs#L204-L211

See attached comment: https://github.com/noir-lang/noir/pull/893#pullrequestreview-1308154733

### PR DESCRIPTION
**Pull request by TomAFrench**
# Description

## Problem\*

Closes #4759 

## Summary\*

I found this due to #6550.

We make heavy use of the `num_bits` method during compilation but this is currently doing a lot of unnecessary allocations in order to perform a full bit decomposition of the field element. We can rework this to work on the byte decomposition which removes 96% of all the temporary allocations from compilation.

## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by asterite** [comment]
Incredible! On Friday I was looking at small and quick optimization opportunities and found a couple of them, one of them was this one (not with this algorithm, but I noticed `num_bits` was allocating too much).

Another thing I found is that `FieldElement::hash` does `state.write(&self.to_be_bytes());` but I was wondering if we could directly call `self.0.hash(state)`, which seems to compile but I don't know why that's not done (though I don't know if it's more efficient).

Another one I found was this double hashmap lookup:

https://github.com/noir-lang/noir/blob/7311d8ca566c3b3e0744389fc5e4163741927767/acvm-repo/acvm/src/compiler/transformers/csat.rs#L274-L275

but I didn't send any PR because I wasn't sure if any of those improved the performance, really.

File: acvm-repo/acir_field/src/field_element.rs [hunk 1]
```diff
@@ -224,12 +207,26 @@ impl<F: PrimeField> AcirField for FieldElement<F> {
...
+
+        // Iterate through the byte decomposition and pop off all leading zeroes
+        let mut iter = bytes.iter().skip_while(|x| (**x) == 0);
+
+        // The first non-zero byte in the decomposition may have some leading zero-bits.
+        let Some(head_byte) = iter.next() else {
+            // If we don't have a non-zero byte then the field element is zero,
+            // which we consider to require a single bit to represent.
+            return 1;
+        };
```
Comment (asterite) [inline acvm-repo/acir_field/src/field_element.rs:220]: Is there a test for this case?

**Review by asterite — APPROVED** [review]
Looks good! I left a comment for something that might require an additional test.

File: acvm-repo/acir_field/src/field_element.rs (hunk 1 above)
Comment (TomAFrench) [reply acvm-repo/acir_field/src/field_element.rs:220]: Just added one.

**Comment by TomAFrench** [comment]
> Another one I found was this double hashmap lookup:

We end up hitting our peak memory usage during SSA rather than in the post-compilation transformations so I think we can ignore this for now.

**Pull request merged**